src/backend.rs

```diff
@@ -0,0 +1,465 @@
+                                        unsafe {
+                                            // libevent dispatches the callback with the mapped file descriptor
+                                            libevent_sys::evmap_io_active_(base.0, fd, libevent_sys::EV_WRITE as i16);
+                                        }
+
+                                        // If the ready flag is not cleared, then this loop will hang the runtime.
+                                        guard.clear_ready();
+                                    }
+                                    Err(error) => {
+                                        tracing::error!(?error);
+                                        break;
+                                    }
+                                }
+                            },
+                            _ = notify.notified() => break, // terminate the task on notification
+                        }
+                    }
+
+                    tracing::debug!(?fd, "I/O task terminated");
+                });
+
+                0
+            }
+            Err(error) => {
+                tracing::error!(?error);
+                -1
+            }
+        }
+    }
+
+    /// Terminates an active I/O task
+    #[instrument]
+    fn del_io(&self, fdinfo: *mut Arc<Notify>) -> c_int {
+        tracing::debug!("delete an I/O event");
+
+        unsafe {
+            match fdinfo.as_mut() {
+                Some(notify) => {
+                    notify.notify_one();
+                    std::ptr::drop_in_place(fdinfo);
+                    0
+                }
+                None => -1,
+            }
+        }
+    }
+
+    /// Drive the tokio runtime with an optional duration for timout events
+    #[instrument]
+    fn dispatch(&self, timeout: Option<Duration>) {
+        self.runtime.block_on(async move {
+            match timeout {
+                // spawned tasks are serviced during the sleep time
+                Some(timeout) => tokio::time::sleep(timeout).await,
+                // at least a single yield is required to advance any pending tasks
+                None => tokio::task::yield_now().await,
+            }
+        })
+    }
+
+    /// Creates a task to service a libevent signal request
+    ///
+    /// A task must continue to provide signal notifications until explicitly
+    /// removed. Note that libevent does not provide user data per signal
+    /// event. Therefore, signals are mapped to notifications in TokioBackend
+    /// to allow for clean task shutdown.
+    ///
+    /// Since the tokio signal handler is installed globally. It is safe to
+    /// handle signals with both libevent and using tokio directly.
+    #[instrument(skip(base))]
+    fn add_signal(
+        &mut self,
+        base: *mut libevent_sys::event_base,
+        nsignal: c_int,
+        events: c_short,
+    ) -> c_int {
+        let base = BaseWrapper(base as *mut libevent_sys::event_base);
```

Comment:
Doesn't look like a cast should be required here on the arg?

Reply:
Thanks. I was passing the base as a reference at one point and missed this change. I don't think `clippy` tries very hard on raw pointer issues.